src/Microsoft.Kusto.ServiceLayer/DataSource/KustoClient.cs

```diff
@@ -187,7 +195,14 @@ public IDataReader ExecuteQuery(string query, CancellationToken cancellationToke
             }
             catch (KustoRequestException exception) when (exception.FailureCode == 401) // Unauthorized
             {
-                throw new DataSourceUnauthorizedException(exception);
+                if (_retryCount <= 0)
+                {
+                    throw;
+                }
+
+                _retryCount--;
```

Comment:
Or just have the parent function call ExecuteQueryInternal once, if the right exception is thrown, call it one more time. It will just make the retry be simple.

For the long term, I think we should just have multiple retry logic on any calls going over the wire to the server. There are many reasons the call could fail and retries with an some waits in the middle will help our code be more reliable for temporary network issues.

Reply:
Checkout the RetryPolicy used in ReliableDataSourceConnection. I think we should use it in all calls we make to Kusto. I think a better retry policy can be done in a separate checkin. Noting it here, as I find retries missing in general. 

For this change, it is not really a retry but refresh of the token. Another thing to consider for token refresh is, why is this only in ExecuteQuery? It should be in all function which call back to Kusto. All calls to Kusto should eventually had a retry, maybe using RetryPolicy, and also handle token refresh if needed.